### ISSUE TITLE
querying total stake and voting powers for past epochs

### ISSUE THREAD

**Issue by egasimus** [namada-net/namada#3810]
In order to independently compute proposal results, we need to fetch voting power for each vote of each past governance proposal. To achieve this, it appears that we need the following data for each past epoch:

- total NAM staked on chain
  - **Expected:** ABCI query `/vp/pos/total_stake/5000` returns total staked amount at epoch 5000
  - **Actual:** ABCI query `/vp/pos/total_stake/5000` returns `0n`
- list of validators
  - **Expected:** ABCI query `/vp/pos/validator/addresses/5900` returns list of validators at epoch 5900
  - **Actual:** ABCI query `/vp/pos/validator/addresses/5900` returns an empty list.
- stake amount of each validator
  - same as above (possibly different epoch count)
- bond amount of each delegator
  - same as above (possibly different epoch count)

As far as I understood from [`epoched.rs`](https://github.com/anoma/namada/blob/b0b41dcfb5631497b01eb2a794561bf5debd57a1/crates/proof_of_stake/src/epoched.rs#L32), **the `Epoched` struct only keeps data for a fixed number of past epochs.** Could you explain why this is the case and whether there is a workaround for fetching historical data from the node?

**Comment by Fraccaman** [comment]
Hi @egasimus! Right now, this is a limitation of namada. We "prune" some of the PoS data each epoch, meaning that the data you are looking for its not easily accessible and requires a more complex mechanism to query it. Unfortunately, this is not a priority for us right now and we won't look into it in the short future. 
As a workaround, you start a fullnode from scratch and while the fullnode is synching, you can connect an indexer.

Ill' leave the issue open and update it as soon as we start working on this.

**Comment by opsecx** [comment]
@Fraccaman having this data available seems like a big net gain for the "indexing community" - any chance it could be given some priority? (not top ofc but also included at some point)

**Comment by brentstone** [comment]
@opsecx I'm not super familiar with the technical details of indexers, but I think I can offer some insight here:

The pruning occurs in the protocol, so no indexer setting would be able to disrupt or change it. However, an indexer or similar service could cache these pieces of data elsewhere while it reads them as they still exist in Namada and then persist them after they've been pruned from the namada DB.

**Comment by McDaan** [comment]
> @Fraccaman having this data available seems like a big net gain for the "indexing community" - any chance it could be given some priority? (not top ofc but also included at some point)

@opsecx this is available on Undexer v4, we'll push out release to main soon

**Comment by opsecx** [comment]
> > @Fraccaman having this data available seems like a big net gain for the "indexing community" - any chance it could be given some priority? (not top ofc but also included at some point)
> 
> 
> 
> @opsecx this is available on Undexer v4, we'll push out release to main soon

Yes, but in a limited fashion, I understand, due to some of these limitations with the protocol?

**Comment by opsecx** [comment]
> @opsecx I'm not super familiar with the technical details of indexers, but I think I can offer some insight here:
> 
> 
> 
> The pruning occurs in the protocol, so no indexer setting would be able to disrupt or change it. However, an indexer or similar service could cache these pieces of data elsewhere while it reads them as they still exist in Namada and then persist them after they've been pruned from the namada DB.

the setting I was referring to is a namada node config parameter. as I understand it that creates a partial archive node for transactions.

**Comment by McDaan** [comment]
> > > @Fraccaman having this data available seems like a big net gain for the "indexing community" - any chance it could be given some priority? (not top ofc but also included at some point)
> > 
> > 
> > @opsecx this is available on Undexer v4, we'll push out release to main soon
> 
> Yes, but in a limited fashion, I understand, due to some of these limitations with the protocol?

I wouldn't say limited, I mean, it address the mentioned limitation of the protocol by being able to fetch data before it gets pruned - it's been made in a way similar to what @brentstone above.

**Comment by opsecx** [comment]
> > > > @Fraccaman having this data available seems like a big net gain for the "indexing community" - any chance it could be given some priority? (not top ofc but also included at some point)
> 
> > > 
> 
> > > 
> 
> > > @opsecx this is available on Undexer v4, we'll push out release to main soon
> 
> > 
> 
> > Yes, but in a limited fashion, I understand, due to some of these limitations with the protocol?
> 
> 
> 
> I wouldn't say limited, I mean, it address the mentioned limitation of the protocol by being able to fetch data before it gets pruned - it's been made in a way similar to what @brentstone said above.

It's limited in the sense that you can't use a general rpc but will need to have the operator use your dockerized node setup. I'm guessing that's so doing a complete full sync will allow the indexer to catch the data while it's still there. I understand the workaround and it's good, but much better if this was addressed on protocol level.(?)

**Comment by brentstone** [comment]
@opsecx from what I remember, the problem with not pruning this data is that it will blow up the size of the DB. I think it's best to keep the DB lean and store other needed data elsewhere. This is unlikely to be changed at protocol level I'd say.

**Comment by McDaan** [comment]
> > I wouldn't say limited, I mean, it address the mentioned limitation of the protocol by being able to fetch data before it gets pruned - it's been made in a way similar to what @brentstone said above.
> 
> It's limited in the sense that you can't use a general rpc but will need to have the operator use your dockerized node setup. I'm guessing that's so doing a complete full sync will allow the indexer to catch the data while it's still there. I understand the workaround and it's good, but much better if this was addressed on protocol level.(?) (ie actually allowing the data to be retained in the node, in a similar fashion to archive of transactions)

We have tried to extend specific values on the source code and run a node with a modified version of Namada software so as to see if we were able to keep the data from a large amount of epochs instead of 2 - it didn't work, so we had to rotate to another potential solution.

**Comment by opsecx** [comment]
> @opsecx from what I remember, the problem with not pruning this data is that it will blow up the size of the DB. I think it's best to keep the DB lean and store other needed data elsewhere. This is unlikely to be changed at protocol level I'd say.

when we are speaking of voting data this should not be a lot of size?
